### DISCUSSION TITLE
how to update or manage mass storage pyboard ?

### DISCUSSION THREAD

**Discussion by patfrench** [micropython/micropython#14281] [General]
Hello,

I have a quick question:

1. The program waits for the user to paste a hex file.
2. I take the HEX file and program a FRAM memory.
3. I delete the HEX file
4. Read the FRAM memory for verication
5. => step  1

```python
# main.py -- put your code here!
from machine import I2C
from pyb import LED
from time import sleep
import os

# Définition des paramètres I2C
I2C_ADDR = 0b01010111  # Adresse I2C du périphérique (0x57)
I2C_FREQ = 400_000    # Fréquence I2C en Hz (400 kHz)
FRAM_SIZE = 8192     # Taille totale de la mémoire FRAM en octets
PROGRAM_DIR = 'program'  # Répertoire où se trouve le fichier HEX

# Initialisation du bus I2C
i2c = I2C(1, freq=I2C_FREQ)

# Initialisation de la LED embarquée
led_red = LED(1) # 1=red, 2=green, 3=yellow, 4=blue
led_green = LED(2) 
led_yellow = LED(3) 
led_blue = LED(4)

# Read FRAM memory
def read_fram(addr, length=1):
    i2c.writeto(I2C_ADDR, addr.to_bytes(2, 'big'))
    return i2c.readfrom(I2C_ADDR, length)

# Write FRAM memory
def write_fram(addr, data):
    i2c.writeto(I2C_ADDR, addr.to_bytes(2, 'big'))
    i2c.writeto(I2C_ADDR, data)
    
# Fonction main
def main():
    while True:     
        # Wait HEX file in directory "program"
        hex_files = [f for f in os.listdir(PROGRAM_DIR) if f.endswith('.hex')]
        if not hex_files:
            # Wait 5 secondes
            led_red.off()
            led_green.off()
            led_yellow.on()
            print("wait")
            sleep(5)
            continue
        
        led_red.on()
        print("detect")
        
        # process
        hex_file = hex_files[0]
        with open(f"{PROGRAM_DIR}/{hex_file}", 'r') as f:
            hex_data = f.read().strip().replace('\n', '').replace(' ', '').encode('utf-8')
        print("open")

        # write FRAM
        for addr in range(0, FRAM_SIZE, 8):
            write_fram(addr, hex_data[addr:addr+8])
        
        # remove HEX file in directory "program"
        os.remove(f"{PROGRAM_DIR}/{hex_file}")
     
        os.sync()  # Force la synchronisation du système de fichiers
        
        # read FRAM memory
        data = bytearray()
        for addr in range(0, FRAM_SIZE, 8):
            data.extend(read_fram(addr, 8))

        led_green.on()

if __name__ == "__main__":
    main()
```

My problem is that in Thonny the HEX file is deleted, but not in Windows Explorer.
and on the second attempt I have memory allocation problems.

```python
Traceback (most recent call last):
  File "<stdin>", line 103, in <module>
  File "<stdin>", line 81, in main
MemoryError: memory allocation failed, allocating 27648 bytes
```

Thank you

**Comment by andrewleech** [comment]
Unfortunately this is the way windows works with mass storage. Windows expects any mass storage device to be like a flash stick that it's in complete control over, there's no way for it to handle a device that's trying to update the storage dynamically.
As such windows always presents its cached view of the data and if that cache gets out of sync with the real data because micropython has changed it then these kinds of errors occur.
I really don't know of any workaround other than resetting the device to force a USB disconnect / reconnect.

**Reply by Josverl** [reply]
I don't think it's windows, with the 'usb drive' changing contents it's a bit like a broken promiseim a flash drive and I'll remember everything ( except now I change ti's file) 

And by default flash drives are not write cached to allow for humans removing the drive without notice.
The two independent writers to the same Fat cause the corruption

**Reply by andrewleech** [reply]
Ah yes true, mass storage operates at the block level, it's the same as having a filesystem mounted in two OS's simultaneously which is just not supported.

I was meaning read cache though not write cache, some blocks are definitely cached for read access but I guess that doesn't really affect the outcome of the write, more as an explanation why windows doesn't "see" that the file was deleted by micropython.

**Comment by patfrench** [comment]
Hi @Josverl, @andrewleech 
Thank you for these explanations.
Unfortunately I understand that I have no solution.
Except the reset button.

thank you
